h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstSpearman.java

```diff
@@ -0,0 +1,180 @@
+package water.rapids.ast.prims.advmath;
+
+import water.DKV;
+import water.MRTask;
+import water.Scope;
+import water.Value;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.rapids.Env;
+import water.rapids.Val;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.vals.ValNum;
+import water.util.FrameUtils;
+
+public class AstSpearman extends AstPrimitive<AstSpearman> {
+  @Override
+  public int nargs() {
+    return 1 + 3; // Frame ID and ID of two numerical vectors to calculate SCC on.
+  }
+
+  @Override
+  public String[] args() {
+    return new String[]{"frame", "first_column", "second_column"};
+  }
+
+  @Override
+  public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+    final Value value = DKV.get(asts[1].str());
+    if (!value.isFrame()) {
+      throw new IllegalArgumentException(String.format("The given key '%s' is not a frame.", asts[1]));
+    }
+
+    final Frame originalUnsortedFrame = value.get(Frame.class);
+    final int vecIdX = originalUnsortedFrame.find(asts[2].exec(env).getStr());
+    final int vecIdY = originalUnsortedFrame.find(asts[3].exec(env).getStr());
+
+    Scope.enter();
+
+    Frame sortedX = new Frame(originalUnsortedFrame.vec(vecIdX).makeCopy());
+    Scope.track(sortedX);
+    Frame sortedY = new Frame(originalUnsortedFrame.vec(vecIdY).makeCopy());
+    Scope.track(sortedY);
+
+    if (!sortedX.vec(0).isCategorical()) {
+      FrameUtils.labelRows(sortedX, "label");
+      sortedX = sortedX.sort(new int[]{0});
+      Scope.track(sortedX);
+
+    }
+
+    if (!sortedY.vec(0).isCategorical()) {
+      FrameUtils.labelRows(sortedY, "label");
+      sortedY = sortedY.sort(new int[]{0});
+      Scope.track(sortedY);
+    }
+
+    assert sortedX.numRows() == sortedY.numRows();
+    final Vec orderX = Vec.makeZero(sortedX.numRows());
+    final Vec orderY = Vec.makeZero(sortedY.numRows());
+
+    final Vec xLabel = sortedX.vec("label") == null ? sortedX.vec(0) : sortedX.vec("label");
+    final Vec xValue = sortedX.vec(0);
+    final Vec yLabel = sortedY.vec("label") == null ? sortedY.vec(0) : sortedY.vec("label");
+    final Vec yValue = sortedY.vec(0);
+    Scope.track(xLabel);
+    Scope.track(yLabel);
+
+    double lastX = Double.NaN;
+    double lastY = Double.NaN;
+    long skippedX = 0;
+    long skippedY = 0;
+    for (int i = 0; i < orderX.length(); i++) {
```

Comment:
We can make this distributed, but it will take two passes instead of one according to my poor knowledge. 

Any advice @michalkurka please ? Thank you.

Reply:
Reading data this way is indeed very inefficient. Vec.Reader should be used instead at the bare minimum.

2-step MRTask seems like a good idea to me.

Reply:
I used Vec Readers and Writers to do the bare minimum. Just in case.

I've been playing with the two-pass algorithm and found some obstacles in the rank calculation - I can't guarantee all the duplicates will be corrected properly in the second pass, unless some formula to calculate the offsets based on first pass (which is per-chunk sort). A trivial solution which would localize the computation is to do up to `n` passes, where `n` is the number of chunks.